module/VuFindSearch/src/VuFindSearch/Backend/Blender/Response/Json/RecordCollection.php

```diff
@@ -0,0 +1,441 @@
+        array $collections,
+        int $limit,
+        int $blockSize,
+        int $totalCount
+    ): array {
+        $this->response = static::$template;
+        $this->response['response']['numFound'] = $totalCount;
+        $this->rewind();
+
+        if (!$collections) {
+            return [];
+        }
+
+        $backendRecords = [];
+        foreach ($collections as $backendId => $collection) {
+            $records = $collection->getRecords();
+            $label = $this->config->Backends[$backendId];
+            foreach ($records as $record) {
+                $record->setSourceIdentifiers(
+                    $record->getSourceIdentifier(),
+                    $backendId
+                );
+                if ($label) {
+                    $record->addLabel($label, 'source');
+                }
+                $backendRecords[$backendId][] = $record;
+            }
+
+            foreach ($collection->getErrors() as $error) {
+                if (is_string($error) && $label) {
+                    $error = [
+                        'msg' => '%%error%% -- %%label%%',
+                        'tokens' => [
+                            '%%error%%' => $error,
+                            '%%label%%' => $label
+                        ],
+                        'translate' => true,
+                        'translateTokens' => true,
+                    ];
+                }
+                $this->addError($error);
+            }
+        }
+
+        $records = [];
+        $backendIds = array_keys($backendRecords);
+        // Filter out unavailable backends from initial results source list:
+        $initialResults = array_values(
+            array_filter(
+                $this->initialResults,
+                function ($s) use ($backendIds) {
+                    return in_array($s, $backendIds);
+                }
+            )
+        );
+        for ($pos = 0; $pos < $limit; $pos++) {
+            $backendId = $this->getBackendAtPosition(
+                $pos,
+                $blockSize,
+                $backendIds,
+                $initialResults
+            );
+            if (!empty($backendRecords[$backendId])) {
+                $this->add(array_shift($backendRecords[$backendId]), false);
+            }
+        }
+
+        $this->response['facet_counts']['facet_fields']
+            = $this->getMergedFacets($collections);
+
+        return $backendRecords;
+    }
+
+    /**
+     * Slice the record collection
+     *
+     * @param int $offset Offset
+     * @param int $limit  Limit
+     *
+     * @return void
+     */
+    public function slice(int $offset, int $limit): void
+    {
+        $this->records = array_slice(
+            $this->records,
+            $offset,
+            $limit
+        );
+    }
+
+    /**
+     * Add an error message
+     *
+     * @param mixed $error Error
+     *
+     * @return void
+     */
+    public function addError($error): void
```

Comment:
Should the error functionality move to a higher level, or perhaps a trait? It seems potentially reusable.

Reply:
Supposedly in most cases the errors are available in the server response, so they don't need to be added. See module/VuFindSearch/src/VuFindSearch/Backend/Primo/Response/RecordCollection.php for an existing case.

Reply:
Ahh, I see. I guess we should leave this alone, but we could make it a trait in future if a second use case appears. :-)